### PR TITLE
Add infrastructure for Cluster Lifecycle API e2e tests

### PR DESCRIPTION
**Pull request by cxbrowne1207**
*Issue #, if available:*

*Description of changes:*
This PR includes some infrastructure needed to write e2e tests for the Cluster Lifecycle API and a vSphere workload cluster e2e test that utilizes it.

*Testing (if applicable):*
Run `TestVSphereMulticlusterWorkloadClusterAPI` e2e test

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/framework/clustervalidator.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+// WithValidation registers a validation to the ClusterValidator that will be run when Validate is called.
+func (c *ClusterValidator) WithValidation(validation clusterValidation, backoffPeriod time.Duration, maxRetries int) {
+	c.validations = append(c.validations, retriableValidation{validation, backoffPeriod, maxRetries})
+}
+
+// WithWorkloadClusterValidations registers a validation set for a management cluster.
+func (c *ClusterValidator) WithWorkloadClusterValidations() {
+	c.WithValidation(validateClusterReady, 5*time.Second, 60)
+	c.WithValidation(validateControlPlaneNodes, 5*time.Second, 60)
```
Comment (taneyland) [inline test/framework/clustervalidator.go:41]: Are we going to add a check for worker nodes?

File: test/framework/workload.go [hunk 2]
```diff
@@ -21,3 +32,71 @@ func (w *WorkloadCluster) DeleteCluster(opts ...CommandOpt) {
...
 	w.deleteCluster(opts...)
 }
+
+// CreateClusterWithKubectl creates a cluster by applying a workload cluster config using the kubectl client.
+func (w *WorkloadCluster) CreateClusterWithKubectl() {
+	ctx := context.Background()
+	w.T.Logf("Applying workload cluster %s spec for creation located at %s", w.ClusterName, w.ClusterConfigLocation)
+	w.applyClusterManifest(ctx)
+	w.waitForKubeconfig(ctx)
+	time.Sleep(2 * time.Minute)
```
Comment (taneyland) [inline test/framework/workload.go:42]: Is this number (2 mins) derived from the average amount of time it takes for everything to be ready?

**Review by taneyland — COMMENTED** [review]
Very nice! 

Do we plan to follow up in future PRs to expand & add more validations?

File: test/framework/cluster.go [hunk 3]
```diff
@@ -41,7 +44,7 @@ import (
 const (
 	defaultClusterConfigFile         = "cluster.yaml"
 	defaultBundleReleaseManifestFile = "bin/local-bundle-release.yaml"
-	defaultEksaBinaryLocation        = "eksctl anywhere"
+	defaultEksaBinaryLocation        = "eksctl-anywhere"
```
[the remark this answers is not in the document]
Comment (cxbrowne1207) [reply test/framework/cluster.go:47]: This is a mistake, will fix. This is an alias that points to my local binary in the repo.

File: test/framework/workload.go (hunk 2 above)
Comment (cxbrowne1207) [reply test/framework/workload.go:42]: This sleep ideally shouldn't be here, but the test was throwing a `connect: network is unreachable` when trying to use it. Working on fixing this by implementing retries where necessary.

File: test/framework/clustervalidator.go (hunk 1 above)
Comment (cxbrowne1207) [reply test/framework/clustervalidator.go:41]: Yes, we will! Currently working on checking for the existence of EKSA objects first, but that is the plan

**Comment by cxbrowne1207** [comment]
/test all

> Very nice!
> 
> Do we plan to follow up in future PRs to expand & add more validations?

If we deem it necessary, yes. More validations can be added in the future

File: pkg/clients/kubernetes/runtimeclient_wb_test.go [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package kubernetes
+
+var NewRuntimeClient = newRuntimeClient
```
[the remark this answers is not in the document]
Comment (cxbrowne1207) [reply pkg/clients/kubernetes/runtimeclient_wb_test.go:3]: newRuntimeClient is needed for the tests, but they it isn't exported (probably shouldn't export it just to test it). This is a way around it, and not having to put the tests in the same package.

**Review by tatlat — COMMENTED** [review]
Nice! /lgtm

File: test/framework/clustervalidator.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+func validateControlPlaneNodes(ctx context.Context, validateOpts clusterOpts) error {
+	cpNodes := &corev1.NodeList{}
+	err := validateOpts.ClusterClient.List(ctx, cpNodes, client.MatchingLabels{"node-role.kubernetes.io/control-plane": ""})
+	if err != nil {
+		return fmt.Errorf("failed to list controlplane nodes %s", err)
+	}
+
+	cpConfig := validateOpts.ClusterSpec.Cluster.Spec.ControlPlaneConfiguration
+	if len(cpNodes.Items) != cpConfig.Count {
+		return fmt.Errorf("control plane node count does not match expected: %v of %v", len(cpNodes.Items), cpConfig.Count)
```
Comment (g-gaston) [inline test/framework/clustervalidator.go:135]: i think we need to use %d here?

File: test/framework/clustervalidator.go [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+	}
+
+	for _, node := range cpNodes.Items {
+		err := validateNodeReady(node)
+		if err != nil {
```
Comment (g-gaston) [inline test/framework/clustervalidator.go:140]: one liner?

File: test/framework/clustervalidator.go [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+		return fmt.Errorf("failed to list controlplane nodes %s", err)
+	}
+
+	cpConfig := validateOpts.ClusterSpec.Cluster.Spec.ControlPlaneConfiguration
+	if len(cpNodes.Items) != cpConfig.Count {
+		return fmt.Errorf("control plane node count does not match expected: %v of %v", len(cpNodes.Items), cpConfig.Count)
+	}
+
+	for _, node := range cpNodes.Items {
+		err := validateNodeReady(node)
```
Comment (g-gaston) [inline test/framework/clustervalidator.go:139]: We are gonna need to add something else here
this validation only checks there is a determined number of nodes and all nodes are ready. But it doesn't check if those nodes are the nodes we expect. For example, we might be expecting 3 nodes, but after an upgrade, this validation will succeed with two old and one new node.

And easy way would be to check the kubernetes version. There are other things we should check, but I think that's the bare minimum to ensure nodes have been rolled out.

File: test/framework/workload.go [hunk 8]
```diff
@@ -21,3 +32,60 @@ func (w *WorkloadCluster) DeleteCluster(opts ...CommandOpt) {
...
 	w.deleteCluster(opts...)
 }
+
+// ApplyClusterManifest uses client-side logic to create/update objects defined in a cluster yaml manifest.
+func (w *WorkloadCluster) ApplyClusterManifest() {
+	ctx := context.Background()
+	w.T.Logf("Applying workload cluster %s spec located at %s", w.ClusterName, w.ClusterConfigLocation)
+	w.applyClusterManifest(ctx)
+	w.waitForKubeconfig(ctx)
+	w.ValidateClusterState()
```
Comment (g-gaston) [inline test/framework/workload.go:42]: I would prefer if we called this separately. IMO this method should just call kubectl apply. If that succeeds, return. And leave the wait and validations for the test to decide.

In general, the framework should offer exported methods for individual operations and leave the orchestration of those operations to the tests.

File: test/framework/workload.go [hunk 9]
```diff
@@ -21,3 +32,60 @@ func (w *WorkloadCluster) DeleteCluster(opts ...CommandOpt) {
...
+}
+
+// DeleteClusterWithKubectl uses client-side logic to delete a cluster.
+func (w *WorkloadCluster) DeleteClusterWithKubectl() {
+	ctx := context.Background()
+	w.T.Logf("Deleting workload cluster %s with kubectl", w.ClusterName)
+	if err := w.KubectlClient.DeleteCluster(ctx, w.managementCluster(), w.cluster()); err != nil {
+		w.T.Fatalf("failed to delete workload cluster config: %s", err)
+	}
+	w.ValidateClusterDelete()
```
Comment (g-gaston) [inline test/framework/workload.go:53]: same comment about orchestration here

File: test/framework/workload.go [hunk 10]
```diff
@@ -21,3 +32,60 @@ func (w *WorkloadCluster) DeleteCluster(opts ...CommandOpt) {
...
+
+func (w *WorkloadCluster) applyClusterManifest(ctx context.Context) {
+	if err := w.KubectlClient.ApplyManifest(ctx, w.ManagementClusterKubeconfigFile(), w.ClusterConfigLocation); err != nil {
+		w.T.Fatalf("Failed to apply workload cluster config: %s", err)
+	}
+}
+
+func (w *WorkloadCluster) waitForKubeconfig(ctx context.Context) {
+	w.T.Logf("Waiting for workload cluster %s kubeconfig to be available", w.ClusterName)
+	err := retrier.Retry(12, 5*time.Second, func() error {
```
Comment (g-gaston) [inline test/framework/workload.go:65]: if it works this fast, cool
but this seems a bit too aggresive. I'm not sure if all providers will be able to spin up a control plane node in a minute.

File: test/framework/workload.go [hunk 11]
```diff
@@ -21,3 +32,60 @@ func (w *WorkloadCluster) DeleteCluster(opts ...CommandOpt) {
...
+		return w.writeKubeconfigToDisk(ctx)
+	})
+	if err != nil {
+		w.T.Fatalf("Failed waiting for cluster kubeconfig: %s", err)
+	}
+}
+
+func (w *WorkloadCluster) writeKubeconfigToDisk(ctx context.Context) error {
+	secret, err := w.KubectlClient.GetSecretFromNamespace(ctx, w.ManagementClusterKubeconfigFile(), fmt.Sprintf("%s-kubeconfig", w.ClusterName), constants.EksaSystemNamespace)
```
Comment (g-gaston) [inline test/framework/workload.go:74]: do we have a method somewhere to get the kubeconfig secret name? maybe even upstream capi has one we can use.

File: test/framework/workload.go [hunk 12]
```diff
@@ -21,3 +32,60 @@ func (w *WorkloadCluster) DeleteCluster(opts ...CommandOpt) {
...
+		return fmt.Errorf("failed to write kubeconfig to disk: %v", err)
+	}
+
+	_, err = writer.Write(filepath.Base(w.kubeconfigFilePath()), kubeconfig, func(op *filewriter.FileOptions) {
+		op.IsTemp = false
+	})
+	if err != nil {
+		return fmt.Errorf("failed to write kubeconfig to disk: %v", err)
+	}
+	return err
```
Comment (g-gaston) [inline test/framework/workload.go:90]: return nil?

File: test/e2e/adot_test.go [hunk 13]
```diff
@@ -124,11 +124,11 @@ func runCuratedPackagesAdotInstallWithUpdate(test *framework.ClusterE2ETest) {
 }
 
 func runCuratedPackagesAdotInstallSimpleFlow(test *framework.ClusterE2ETest) {
-	test.WithCluster(runCuratedPackagesAdotInstall)
+	test.RunWithCluster(runCuratedPackagesAdotInstall)
```
[the remark this answers is not in the document]
Comment (cxbrowne1207) [reply test/e2e/adot_test.go]: Indeed. This was already fixed

[the remark this answers is not in the document]
Comment (cxbrowne1207) [reply test/e2e/adot_test.go]: Indeed. This was already fixed

File: test/framework/cluster.go [hunk 14]
```diff
@@ -1658,3 +1684,94 @@ func (e *ClusterE2ETest) CombinedAutoScalerMetricServerTest(autoscalerName strin
...
+
+// ValidateClusterDelete verifies the cluster has been deleted.
+func (e *ClusterE2ETest) ValidateClusterDelete() {
+	ctx := context.Background()
+	e.T.Logf("Validating cluster deletion %s", e.ClusterName)
+	if e.clusterValidator == nil {
+		return
+	}
+
+	e.clusterValidator.Reset()
```
[the remark this answers is not in the document]
Comment (cxbrowne1207) [reply test/framework/cluster.go:1715]: The reason for this was creating a new validator requires the a client, but after running the kubectl delete, it was not possible to instantiate a new client, which the code currently does every-time. I opted to re-use the validator, but I probably could have also re-used the client instead.

**Pull request merged**